File: src/other_questions/quick_sort.py
# quick sort


def quick_sort(arr, low, high):
    """
    quick sort
    time complexity: O(NlogN)
    :param arr: array
    :param low: start index
    :param high: end index
    """
    if low < high:
        pivot = arr[high]

        j = low - 1  # smaller element index
        i = low  # loop index
        while i < high:
            if arr[i] < pivot:
                j += 1
                arr[i], arr[j] = arr[j], arr[i]
            i += 1

        # swap the pivot to the correct position
        arr[j + 1], arr[high] = arr[high], arr[j + 1]

        quick_sort(arr, low, j)
        quick_sort(arr, j + 1, high)


if __name__ == "__main__":
    arr = [6, 5, 7, 0, 1, 8, 2, 9, 3, 4]
    quick_sort(arr, 0, len(arr) - 1)
    print(arr)

File: classical-algorithms/quick-sort.go
package main

import "fmt"

func quickSort(arr []int, start int, end int) {
	if start >= end {
		return
	}

	pivot := arr[start]

	low := start + 1
	high := end

	// correct position of the pivot in the end
	pivotEnds := 0
	for low <= high {
		if arr[low] > pivot {
			arr[low], arr[high] = arr[high], arr[low]
			high--
			pivotEnds = high
		} else {
			low++
			pivotEnds = high - 1
		}
	}

	// swap the pivot to correct position
	arr[start], arr[pivotEnds] = arr[pivotEnds], arr[start]

	quickSort(arr, start, high-1)
	quickSort(arr, high+1, end)
}

func main() {
	// arr := []int{5, 1, 2, 4, 9, 3, 7, 8, 0}
	// arr := []int{3,5,3,5}
	arr := []int{2,2,1,1,1,1,2}

	quickSort(arr, 0, len(arr)-1)
	fmt.Println(arr)
}
